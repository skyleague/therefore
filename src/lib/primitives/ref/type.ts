/* eslint-disable @typescript-eslint/no-unsafe-declaration-merging */
import type { NodeTrait } from '../../cst/mixin.js'
import { Node } from '../../cst/node.js'
import type { Intrinsic } from '../../cst/types.js'
import type { SchemaOptions } from '../base.js'

import { type ConstExpr, evaluate } from '@skyleague/axioms'

export type RefOptions = object

export class RefType<Reference extends Node = Node> extends Node {
    public override _type = 'ref' as const
    public override _children: [Node]

    public _options: RefOptions = {}
    public declare infer: Reference['infer']
    public declare input: Reference['input']
    public declare intrinsic: Intrinsic<Reference>

    public constructor(reference: ConstExpr<Reference>, options: SchemaOptions<RefOptions, Reference['infer']>) {
        super(options)
        this._options = options

        // we are lying to typescript here, but it's fine
        // there expression will be evaluated in the prepass, and errors hard if it's not
        this._children = [reference as Node]
    }

    public override _hooks = {
        onLoad: [
            (self: Node): void => {
                self._children = self._children?.map(evaluate) as [Node]
            },
        ],
        onGenerate: [
            ({ _children: children }: Node): void => {
                const [reference] = children ?? []
                if (reference !== undefined && reference._sourcePath === undefined) {
                    // this reference was not exported on its own
                    // so we just call it local if has no name
                    reference._name ??= 'local'
                }
            },
        ],
    }

    public get item(): Reference {
        return this._children[0] as Reference
    }

    public static _from<Reference extends Node>(
        ref: ConstExpr<Reference>,
        options: SchemaOptions<RefOptions>,
    ): RefType<Intrinsic<Reference>> {
        type IntrinsicType = Intrinsic<Reference>
        return new RefType<IntrinsicType>(ref as ConstExpr<IntrinsicType>, options)
    }
}

export interface RefType extends Node, NodeTrait {}
