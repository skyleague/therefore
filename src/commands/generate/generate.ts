import { generatedAjv, generatedBy } from './constants.js'
import { expandGlobs } from './glob.js'
import { formatFile, maybeLoadPrettier } from './prettier.js'
import { resolveTypescriptSchema } from './resolver.js'
import type { FileDefinition, OutputFile, ReferenceData, ThereforeOutputType } from './types.js'

import { getExtension, replaceExtension } from '../../common/template/path.js'
import { renderTemplate } from '../../common/template/template.js'
import type { ThereforeNode } from '../../lib/cst/cst.js'
import type { ThereforeCst } from '../../lib/primitives/types.js'
import { isThereforeExport } from '../../lib/primitives/types.js'
import { toJsonSchema } from '../../lib/visitor/jsonschema/jsonschema.js'
import { prepass } from '../../lib/visitor/prepass/prepass.js'
import { toTypescriptDefinition } from '../../lib/visitor/typescript/typescript.js'

import { entriesOf, enumerate, groupBy, isDefined, range, second } from '@skyleague/axioms'
import decamelize from 'decamelize'

import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
async function requireModule(module: string): Promise<Record<string, ThereforeNode | unknown>> {
    const relative = path.relative(__dirname, module).replace(/\\/g, '/').replace(/\.ts$/, '.js')

    const mod = (await import(relative.startsWith('.') ? relative : `./${relative}`)) as Record<string, unknown>
    return (mod.default ?? mod) as Record<string, ThereforeNode | unknown>
}

export async function loadSymbol({
    srcPath,
    basePath,
    symbol,
    sourceSymbol,
    compile,
    definitions,
    entry,
    fileSuffix,
    filePath,
    outputFileRename,
    appendSourceSymbolName: appendSourceName = false,
    rootPath = srcPath,
}: {
    srcPath: string
    basePath: string
    symbol: ThereforeCst
    sourceSymbol: string
    compile: boolean
    definitions: Record<string, FileDefinition | undefined>
    entry: string
    fileSuffix?: string | undefined
    filePath?: string | undefined
    outputFileRename: (path: string) => string
    appendSourceSymbolName?: boolean
    rootPath?: string
}) {
    const simplified = prepass(symbol)
    let targetDir = '.'
    if (simplified.type === 'custom') {
        if (simplified.value.fileSuffix !== undefined) {
            fileSuffix = simplified.value.fileSuffix
        }
        if (simplified.value.filePath !== undefined) {
            const dir = path.dirname(simplified.value.filePath)
            const basename = path.basename(simplified.value.filePath)
            filePath = basename
            targetDir = dir
        }
    }

    const { base: baseName, dir: baseDir } = path.parse(srcPath)
    const ext = getExtension(srcPath) ?? ''
    const targetBaseDir = path.relative(basePath, path.join(baseDir, targetDir))
    const targetSrcPath = `${targetBaseDir}/${(filePath ?? baseName).replace(ext, '')}${ext}`
    const targetPath = fileSuffix !== undefined ? replaceExtension(targetSrcPath, fileSuffix) : outputFileRename(targetSrcPath)

    definitions[targetPath] ??= {
        srcPath: path.relative(basePath, rootPath),
        targetPath,
        attachedFiles: [],
        symbols: [],
        dependencies: {},
    }

    const file = definitions[targetPath]!
    // check if symbol is not seen yet
    if (file.symbols.find((s) => s.uuid === simplified.uuid)) {
        return
    }
    const symbolName = simplified.name ?? sourceSymbol

    sourceSymbol = appendSourceName && simplified.name !== undefined ? `${sourceSymbol}.${simplified.name}` : symbolName
    const schemaName = decamelize(symbolName, { separator: '-' })
    const schemaFile = `./schemas/${schemaName}.schema.json`
    const compiledFile = `./schemas/${schemaName}.schema.js`

    const nodeIsCompiled = simplified.description.validator?.compile ?? compile

    const jsonschema = await toJsonSchema(simplified, nodeIsCompiled)
    const { definition, subtrees } = toTypescriptDefinition({ sourceSymbol, symbolName, schema: simplified })

    file.symbols.push(
        ...Object.values(definition.locals ?? {})
            .filter(isDefined)
            .map((local) => ({
                uuid: local.uuid,
                symbolName: local.symbolName,
                definition: local,
                typeOnly: true,
            }))
    )

    file.symbols.push({
        uuid: simplified.uuid,
        symbolName,
        definition: definition,
        schemaFile,
        compiledFile: nodeIsCompiled ? compiledFile : undefined,
        typeOnly: simplified.description.validator?.enabled !== true,
    })

    if (definition.isExported && simplified.description.validator?.enabled) {
        const cleanSchemasFolder = () => {
            const targetFolder = path.dirname(path.join(targetBaseDir, compiledFile))
            if (fs.existsSync(targetFolder)) {
                console.warn(`Cleaning ${targetFolder}`)
                fs.rmSync(targetFolder, { force: true, recursive: true })
            }
        }
        if (jsonschema.compiled) {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, compiledFile),
                content: `/**\n * ${generatedAjv} \n * eslint-disable\n */\n${jsonschema.code}`,
                prettify: false,
                type: 'validator',
                clean: cleanSchemasFolder,
            })
        } else {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, schemaFile),
                content: JSON.stringify(jsonschema.schema, null, 2),
                prettify: true,
                type: 'jsonschema',
                clean: cleanSchemasFolder,
            })
        }

        console.debug(` - found ${definition.symbolName}`)
    }
    for (const { node: subSymbol, fileSuffix: subFileSuffix, filePath: subFilePath } of subtrees) {
        await loadSymbol({
            symbol: subSymbol,
            basePath,
            sourceSymbol,
            compile,
            definitions,
            srcPath: path.join(basePath, targetPath),
            rootPath: srcPath,
            fileSuffix: subFileSuffix,
            filePath: subFilePath,
            entry,
            outputFileRename,
            appendSourceSymbolName: true,
        })
    }
}

export async function scanModule({
    entry,
    srcPath,
    basePath,
    compile,
    definitions,
    require = requireModule,
    outputFileRename,
}: {
    entry: string
    srcPath: string
    basePath: string
    compile: boolean
    definitions: Record<string, FileDefinition>
    require?: (module: string) => Promise<Record<string, ThereforeNode | unknown>>
    outputFileRename: (path: string) => string
}) {
    const module = await require(entry)

    for (const [sourceSymbol, symbolPromise] of Object.entries(module)) {
        const symbol = await symbolPromise

        if (!isThereforeExport(symbol)) {
            continue
        }

        await loadSymbol({ symbol, sourceSymbol, compile, definitions, srcPath, entry, basePath, outputFileRename })
    }
}

export async function scanFiles({
    files,
    basePath,
    compile,
    outputFileRename,
}: {
    files: string[]
    basePath: string
    compile: boolean
    outputFileRename: (path: string) => string
}): Promise<Record<string, FileDefinition>> {
    const definitions: Record<string, FileDefinition> = {}
    for (const entry of files) {
        const srcPath = path.resolve(basePath, entry)
        let isGenerated = false
        const strippedPath = srcPath.substr(0, srcPath.lastIndexOf('.'))
        for (const ext of ['.ts', '.js']) {
            const p = `${strippedPath}${ext}`
            isGenerated ||= fs.existsSync(p) && fs.readFileSync(p).includes(generatedBy)
        }
        if (isGenerated) {
            console.debug(`scanning ${entry}`)
            console.debug(` * skipping generated schema`)
            continue
        }

        console.log(`scanning ${entry}`)
        try {
            await scanModule({ entry, basePath, srcPath, compile, definitions, outputFileRename })
        } catch (e: unknown) {
            const error = e as Error
            console.debug(error.message, error.stack)
            throw error
        }
    }
    return definitions
}

export async function compileOutputFiles(
    entries: string[],
    { compile = false, cwd, outputFileRename }: { cwd: string; compile: boolean; outputFileRename: (path: string) => string }
): Promise<
    {
        targetPath: string
        template: string
        data: Record<string, ReferenceData>
        type: ThereforeOutputType
        prettify: boolean
        clean?: () => void
    }[]
> {
    const outputFiles: OutputFile[] = []
    const definitions = await scanFiles({ files: entries, basePath: cwd, compile, outputFileRename })

    const localReferences = {}

    for (const targetPath of Object.keys(definitions)) {
        resolveTypescriptSchema({ definitions, targetPath, cwd, outputFiles, localReferences })
    }

    return compile ? outputFiles : outputFiles.filter((f) => f.type !== 'validator')
}

export async function generate({
    globs,
    ignore,
    extension,
    compile,
    outputFileRename,
    clean,
}: {
    globs: string[]
    ignore: string[]
    extension: string
    compile: boolean
    outputFileRename: (path: string) => string
    clean: boolean
}): Promise<void> {
    const cwd = process.cwd()
    const entries = await expandGlobs({ patterns: globs, ignore: [`!${extension}`, ...ignore], cwd, extension })

    const outputFiles = await compileOutputFiles(entries, {
        cwd,
        outputFileRename,
        compile,
    })

    if (clean) {
        for (const of of outputFiles) {
            if (of.clean !== undefined) {
                of.clean()
            }
        }
    }

    const prettier = maybeLoadPrettier()

    const data = outputFiles.reduce<Record<string, ReferenceData>>(
        (agg, x) => ({
            ...agg,
            ...x.data,
        }),
        {}
    )

    const duplicates = entriesOf(groupBy(entriesOf(data), ([, values]) => `${values.srcPath}${values.symbolName}`)).filter(
        ([, records]) => records.length > 1
    )
    for (const duplicateGroup of duplicates.map((d) => second(d))) {
        for (const [i, [key]] of enumerate(duplicateGroup)) {
            // data[key].referenceName = `${data[key].referenceName}${i}`
            data[key]!.symbolName = `${data[key]!.symbolName}${i}`
            data[key]!.uniqueSymbolName = `${data[key]!.uniqueSymbolName}${i}`
        }
    }

    const variablesMap = Object.fromEntries(
        Object.entries(data)
            .map(([key, vals]) => Object.entries(vals).map(([varName, value]) => [`${key}:${varName}`, value] as const))
            .flat()
    )

    for (const { targetPath, template, type, prettify } of outputFiles) {
        fs.mkdirSync(path.dirname(targetPath), { recursive: true })
        // allow nested references up to 3 deep
        let contents = template
        for (const _ of range(3)) {
            contents = renderTemplate(contents, variablesMap)
        }
        fs.writeFileSync(targetPath, prettify ? await formatFile(prettier, contents, targetPath, type) : contents)
    }
}
