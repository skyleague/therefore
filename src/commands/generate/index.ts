import { generate } from './generate.js'

import { constants } from '../../lib/constants.js'

import type { Argv } from 'yargs'

export function builder(yargs: Argv) {
    return yargs
        .option('files', {
            alias: 'f',
            demandOption: true,
            describe: 'globs to scan for schemas',
            type: 'array',
        })
        .option('ignore-pattern', {
            alias: 'i',
            demandOption: true,
            default: ['**/*.d.ts', 'node_modules'],
            describe: 'globs to exclude',
            type: 'array',
        })
        .option('clean', {
            default: true,
            type: 'boolean',
        })
        .option('migrate-to', {
            describe: 'migrate schemas to specified type',
            type: 'string',
            choices: ['zod'],
            default: undefined,
        })
}

export async function handler(argv: ReturnType<typeof builder>['argv']): Promise<void> {
    const { files = [], 'ignore-pattern': ignorePatterns, clean, 'migrate-to': migrateTo } = await argv

    if (migrateTo) {
        constants.migrateToValidator = migrateTo as 'zod' | 'ajv'
        constants.migrate = true
    }
    await generate({
        globs: files.map((f) => f.toString()),
        ignore: ignorePatterns.map((p) => p.toString()),
        extension: constants.schemaExtension,
        clean,
    })
}

export default {
    command: ['generate', '*'],
    describe: 'generate therefore types',
    builder,
    handler,
}
