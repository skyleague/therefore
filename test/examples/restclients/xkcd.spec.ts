import { XKCD } from '../../../examples/restclients/xkcd/xkcd.client.js'
import type { Comic } from '../../../examples/restclients/xkcd/xkcd.type.js'
import { compileOutput } from '../../../src/commands/generate/generate.js'

import { describe, expect, it } from 'vitest'

it(' output generation', async () => {
    expect(
        await compileOutput(['examples/restclients/xkcd/xkcd.schema.ts'], {
            cwd: process.cwd(),
        }),
    ).toMatchSnapshot()
})

describe('xkcd', () => {
    const client = new XKCD()
    it('simple get', async () => {
        const result = await client.getInfo0JsonByComicId({ path: { comicId: '2653' } })
        const comic: Comic = result.body
        void comic
        expect(result).toMatchInlineSnapshot(
            {
                headers: expect.any(Object),
            },
            `
          {
            "body": {
              "alt": ""My parents were both omnitaurs, which is how I got interested in recombination," said the normal human.",
              "day": "1",
              "img": "https://imgs.xkcd.com/comics/omnitaur.png",
              "link": "",
              "month": "8",
              "news": "",
              "num": 2653,
              "safe_title": "Omnitaur",
              "title": "Omnitaur",
              "transcript": "",
              "year": "2022",
            },
            "headers": Any<Object>,
            "statusCode": 200,
          }
        `,
        )
    })
})
