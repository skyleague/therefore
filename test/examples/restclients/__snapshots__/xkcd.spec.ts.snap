// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`output generation - ajv 1`] = `
{
  "cleanFn": [Function],
  "outputFiles": {
    "examples/restclients/ajv/xkcd/nullable/nullable.client.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import { got } from 'got'
import type { CancelableRequest, Got, Options, OptionsInit, Response } from 'got'

import { Comic } from './nullable.type.js'

/**
 * XKCD
 * 
 * Webcomic of romance, sarcasm, math, and language.
 */
export class XKCDNullable {
    public client: Got

    public constructor(
    {
        prefixUrl = 'http://xkcd.com/',
        options,
        client = got
    }: {
        prefixUrl?: string | 'http://xkcd.com/',
        options?: Options | OptionsInit,
        client?: Got
    } = {}) {
        this.client = client.extend(...[{ prefixUrl }, options].filter((o): o is Options => o !== undefined))
    }


    /**
     * GET /info.0.json
     * 
     * Fetch current comic and metadata.
     */
    public getInfo0Json()
    {
        return this.awaitResponse(this.client.get('info.0.json', {
            responseType: 'json',
        }), {
            200: Comic,
        })
    }

    /**
     * GET /{comicId}/info.0.json
     * 
     * Fetch comics and metadata  by comic id.
     */
    public getInfo0JsonByComicId({path}: {path: { comicId: string }})
    {
        return this.awaitResponse(this.client.get(\`\${path.comicId}/info.0.json\`, {
            responseType: 'json',
        }), {
            200: Comic,
        })
    }

    public async awaitResponse<T, S extends Record<PropertyKey, undefined | { is: (o: unknown) => o is T; assert?: (o: unknown) => void }>>(response: 
    CancelableRequest<Response>
    , schemas: S) {
        type FilterStartingWith<S extends PropertyKey, T extends string> = S extends number | string ? \`\${S}\` extends \`\${T}\${infer _X}\` ? S : never : never
        type InferSchemaType<T> = T extends { is: (o: unknown) => o is infer S } ? S : never
        const result = await response
        const schema = schemas[result.statusCode] ?? schemas.default
        schema?.assert?.(result.body)
        return {statusCode: result.statusCode, headers: result.headers, body: result.body as InferSchemaType<S[keyof Pick<S, FilterStartingWith<keyof S, '2' | 'default'>>]> }
    }
}

",
    "examples/restclients/ajv/xkcd/nullable/nullable.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import { Ajv, ValidationError } from 'ajv'
import type { DefinedError } from 'ajv'

import ComicSchema from './schemas/comic.schema.json' with { type: "json" }

export interface Comic {
    alt?: (string | null | undefined)
    day?: (string | null | undefined)
    img?: (string | null | undefined)
    link?: (string | null | undefined)
    month?: (string | null | undefined)
    news?: (string | null | undefined)
    num?: (number | null | undefined)
    safe_title?: (string | null | undefined)
    title?: (string | null | undefined)
    transcript?: (string | null | undefined)
    year?: (string | null | undefined)
}

export const Comic = {
    validate: new Ajv({"strict":true,"strictSchema":false,"strictTypes":true,"strictTuples":false,"useDefaults":true,"logger":false,"loopRequired":5,"loopEnum":5,"multipleOfPrecision":4,"code":{"esm":true}}).compile<Comic>(ComicSchema),
    schema: ComicSchema,
    get errors() { return Comic.validate.errors ?? undefined },
    is: (o: unknown): o is Comic => Comic.validate(o) === true,
    assert: (o: unknown) => {
        if (!Comic.validate(o)) {
            throw new ValidationError(Comic.errors ?? [])
        }
    },
    parse: (o: unknown): { right: Comic } | { left: DefinedError[] } => 
    {
        if(Comic.is(o)) {
            return { right: o }
        }
        return { left: (Comic.errors ?? []) as DefinedError[] }
    },
} as const


",
    "examples/restclients/ajv/xkcd/nullable/schemas/comic.schema.json": "{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Comic",
  "type": "object",
  "properties": {
    "alt": {
      "type": [
        "string",
        "null"
      ]
    },
    "day": {
      "type": [
        "string",
        "null"
      ]
    },
    "img": {
      "type": [
        "string",
        "null"
      ]
    },
    "link": {
      "type": [
        "string",
        "null"
      ]
    },
    "month": {
      "type": [
        "string",
        "null"
      ]
    },
    "news": {
      "type": [
        "string",
        "null"
      ]
    },
    "num": {
      "type": [
        "number",
        "null"
      ]
    },
    "safe_title": {
      "type": [
        "string",
        "null"
      ]
    },
    "title": {
      "type": [
        "string",
        "null"
      ]
    },
    "transcript": {
      "type": [
        "string",
        "null"
      ]
    },
    "year": {
      "type": [
        "string",
        "null"
      ]
    }
  },
  "additionalProperties": true
}",
    "examples/restclients/ajv/xkcd/schemas/comic.schema.js": "/* eslint-disable */
// @ts-nocheck
/**
 * Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code
 */
"use strict";
/** @type {unknown} */
export const validate = validate10;export default validate10;const schema11 = {"$schema":"http://json-schema.org/draft-07/schema#","title":"Comic","type":"object","properties":{"alt":{"type":"string"},"day":{"type":"string"},"img":{"type":"string"},"link":{"type":"string"},"month":{"type":"string"},"news":{"type":"string"},"num":{"type":"number"},"safe_title":{"type":"string"},"title":{"type":"string"},"transcript":{"type":"string"},"year":{"type":"string"}},"additionalProperties":true};function validate10(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(errors === 0){if(data && typeof data == "object" && !Array.isArray(data)){if(data.alt !== undefined){const _errs2 = errors;if(typeof data.alt !== "string"){validate10.errors = [{instancePath:instancePath+"/alt",schemaPath:"#/properties/alt/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs2 === errors;}else {var valid0 = true;}if(valid0){if(data.day !== undefined){const _errs4 = errors;if(typeof data.day !== "string"){validate10.errors = [{instancePath:instancePath+"/day",schemaPath:"#/properties/day/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs4 === errors;}else {var valid0 = true;}if(valid0){if(data.img !== undefined){const _errs6 = errors;if(typeof data.img !== "string"){validate10.errors = [{instancePath:instancePath+"/img",schemaPath:"#/properties/img/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs6 === errors;}else {var valid0 = true;}if(valid0){if(data.link !== undefined){const _errs8 = errors;if(typeof data.link !== "string"){validate10.errors = [{instancePath:instancePath+"/link",schemaPath:"#/properties/link/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs8 === errors;}else {var valid0 = true;}if(valid0){if(data.month !== undefined){const _errs10 = errors;if(typeof data.month !== "string"){validate10.errors = [{instancePath:instancePath+"/month",schemaPath:"#/properties/month/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs10 === errors;}else {var valid0 = true;}if(valid0){if(data.news !== undefined){const _errs12 = errors;if(typeof data.news !== "string"){validate10.errors = [{instancePath:instancePath+"/news",schemaPath:"#/properties/news/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs12 === errors;}else {var valid0 = true;}if(valid0){if(data.num !== undefined){let data6 = data.num;const _errs14 = errors;if(!((typeof data6 == "number") && (isFinite(data6)))){validate10.errors = [{instancePath:instancePath+"/num",schemaPath:"#/properties/num/type",keyword:"type",params:{type: "number"},message:"must be number"}];return false;}var valid0 = _errs14 === errors;}else {var valid0 = true;}if(valid0){if(data.safe_title !== undefined){const _errs16 = errors;if(typeof data.safe_title !== "string"){validate10.errors = [{instancePath:instancePath+"/safe_title",schemaPath:"#/properties/safe_title/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs16 === errors;}else {var valid0 = true;}if(valid0){if(data.title !== undefined){const _errs18 = errors;if(typeof data.title !== "string"){validate10.errors = [{instancePath:instancePath+"/title",schemaPath:"#/properties/title/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs18 === errors;}else {var valid0 = true;}if(valid0){if(data.transcript !== undefined){const _errs20 = errors;if(typeof data.transcript !== "string"){validate10.errors = [{instancePath:instancePath+"/transcript",schemaPath:"#/properties/transcript/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs20 === errors;}else {var valid0 = true;}if(valid0){if(data.year !== undefined){const _errs22 = errors;if(typeof data.year !== "string"){validate10.errors = [{instancePath:instancePath+"/year",schemaPath:"#/properties/year/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs22 === errors;}else {var valid0 = true;}}}}}}}}}}}}else {validate10.errors = [{instancePath,schemaPath:"#/type",keyword:"type",params:{type: "object"},message:"must be object"}];return false;}}validate10.errors = vErrors;return errors === 0;};validate.schema=schema11;",
    "examples/restclients/ajv/xkcd/xkcd.client.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import { got } from 'got'
import type { CancelableRequest, Got, Options, OptionsInit, Response } from 'got'

import { Comic } from './xkcd.type.js'

/**
 * XKCD
 * 
 * Webcomic of romance, sarcasm, math, and language.
 */
export class XKCD {
    public client: Got

    public constructor(
    {
        prefixUrl = 'http://xkcd.com/',
        options,
        client = got
    }: {
        prefixUrl?: string | 'http://xkcd.com/',
        options?: Options | OptionsInit,
        client?: Got
    } = {}) {
        this.client = client.extend(...[{ prefixUrl }, options].filter((o): o is Options => o !== undefined))
    }


    /**
     * GET /info.0.json
     * 
     * Fetch current comic and metadata.
     */
    public getInfo0Json()
    {
        return this.awaitResponse(this.client.get('info.0.json', {
            responseType: 'json',
        }), {
            200: Comic,
        })
    }

    /**
     * GET /{comicId}/info.0.json
     * 
     * Fetch comics and metadata  by comic id.
     */
    public getInfo0JsonByComicId({path}: {path: { comicId: string }})
    {
        return this.awaitResponse(this.client.get(\`\${path.comicId}/info.0.json\`, {
            responseType: 'json',
        }), {
            200: Comic,
        })
    }

    public async awaitResponse<T, S extends Record<PropertyKey, undefined | { is: (o: unknown) => o is T; assert?: (o: unknown) => void }>>(response: 
    CancelableRequest<Response>
    , schemas: S) {
        type FilterStartingWith<S extends PropertyKey, T extends string> = S extends number | string ? \`\${S}\` extends \`\${T}\${infer _X}\` ? S : never : never
        type InferSchemaType<T> = T extends { is: (o: unknown) => o is infer S } ? S : never
        const result = await response
        const schema = schemas[result.statusCode] ?? schemas.default
        schema?.assert?.(result.body)
        return {statusCode: result.statusCode, headers: result.headers, body: result.body as InferSchemaType<S[keyof Pick<S, FilterStartingWith<keyof S, '2' | 'default'>>]> }
    }
}

",
    "examples/restclients/ajv/xkcd/xkcd.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import { ValidationError } from 'ajv'
import type { DefinedError, ValidateFunction } from 'ajv'

import { validate as ComicValidator } from './schemas/comic.schema.js'

export interface Comic {
    alt?: (string | undefined)
    day?: (string | undefined)
    img?: (string | undefined)
    link?: (string | undefined)
    month?: (string | undefined)
    news?: (string | undefined)
    num?: (number | undefined)
    safe_title?: (string | undefined)
    title?: (string | undefined)
    transcript?: (string | undefined)
    year?: (string | undefined)
}

export const Comic = {
    validate: ComicValidator as ValidateFunction<Comic>,
    get schema() { return Comic.validate.schema},
    get errors() { return Comic.validate.errors ?? undefined },
    is: (o: unknown): o is Comic => Comic.validate(o) === true,
    assert: (o: unknown) => {
        if (!Comic.validate(o)) {
            throw new ValidationError(Comic.errors ?? [])
        }
    },
    parse: (o: unknown): { right: Comic } | { left: DefinedError[] } => 
    {
        if(Comic.is(o)) {
            return { right: o }
        }
        return { left: (Comic.errors ?? []) as DefinedError[] }
    },
} as const


",
  },
}
`;

exports[`output generation - zod 1`] = `
{
  "cleanFn": [Function],
  "outputFiles": {
    "examples/restclients/zod/xkcd/nullable/nullable.client.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import { got } from 'got'
import type { CancelableRequest, Got, Options, OptionsInit, Response } from 'got'

import { Comic } from './nullable.zod.js'

/**
 * XKCD
 * 
 * Webcomic of romance, sarcasm, math, and language.
 */
export class XKCDNullable {
    public client: Got

    public constructor(
    {
        prefixUrl = 'http://xkcd.com/',
        options,
        client = got
    }: {
        prefixUrl?: string | 'http://xkcd.com/',
        options?: Options | OptionsInit,
        client?: Got
    } = {}) {
        this.client = client.extend(...[{ prefixUrl }, options].filter((o): o is Options => o !== undefined))
    }


    /**
     * GET /info.0.json
     * 
     * Fetch current comic and metadata.
     */
    public getInfo0Json()
    {
        return this.awaitResponse(this.client.get('info.0.json', {
            responseType: 'json',
        }), {
            200: Comic,
        })
    }

    /**
     * GET /{comicId}/info.0.json
     * 
     * Fetch comics and metadata  by comic id.
     */
    public getInfo0JsonByComicId({path}: {path: { comicId: string }})
    {
        return this.awaitResponse(this.client.get(\`\${path.comicId}/info.0.json\`, {
            responseType: 'json',
        }), {
            200: Comic,
        })
    }

    public async awaitResponse<I, S extends Record<PropertyKey, { parse: (o: unknown) => I }>>(response: 
    CancelableRequest<Response>
    , schemas: S) {
        type FilterStartingWith<S extends PropertyKey, T extends string> = S extends number | string ? \`\${S}\` extends \`\${T}\${infer _X}\` ? S : never : never
        type InferSchemaType<T> = T extends { is: (o: unknown) => o is infer S } ? S : never
        const result = await response
        const schema = schemas[result.statusCode] ?? schemas.default
        const body = schema?.parse?.(result.body) ?? result.body
        return {statusCode: result.statusCode, headers: result.headers, body: body as InferSchemaType<S[keyof Pick<S, FilterStartingWith<keyof S, '2' | 'default'>>]> }
    }
}

",
    "examples/restclients/zod/xkcd/nullable/nullable.zod.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import { z } from 'zod'

export const Comic = z.object({
    alt: z.string().nullable().optional(),
    day: z.string().nullable().optional(),
    img: z.string().nullable().optional(),
    link: z.string().nullable().optional(),
    month: z.string().nullable().optional(),
    news: z.string().nullable().optional(),
    num: z.number().nullable().optional(),
    safe_title: z.string().nullable().optional(),
    title: z.string().nullable().optional(),
    transcript: z.string().nullable().optional(),
    year: z.string().nullable().optional(),
})

export type Comic = z.infer<typeof Comic>

",
    "examples/restclients/zod/xkcd/xkcd.client.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import { got } from 'got'
import type { CancelableRequest, Got, Options, OptionsInit, Response } from 'got'

import { Comic } from './xkcd.zod.js'

/**
 * XKCD
 * 
 * Webcomic of romance, sarcasm, math, and language.
 */
export class XKCD {
    public client: Got

    public constructor(
    {
        prefixUrl = 'http://xkcd.com/',
        options,
        client = got
    }: {
        prefixUrl?: string | 'http://xkcd.com/',
        options?: Options | OptionsInit,
        client?: Got
    } = {}) {
        this.client = client.extend(...[{ prefixUrl }, options].filter((o): o is Options => o !== undefined))
    }


    /**
     * GET /info.0.json
     * 
     * Fetch current comic and metadata.
     */
    public getInfo0Json()
    {
        return this.awaitResponse(this.client.get('info.0.json', {
            responseType: 'json',
        }), {
            200: Comic,
        })
    }

    /**
     * GET /{comicId}/info.0.json
     * 
     * Fetch comics and metadata  by comic id.
     */
    public getInfo0JsonByComicId({path}: {path: { comicId: string }})
    {
        return this.awaitResponse(this.client.get(\`\${path.comicId}/info.0.json\`, {
            responseType: 'json',
        }), {
            200: Comic,
        })
    }

    public async awaitResponse<I, S extends Record<PropertyKey, { parse: (o: unknown) => I }>>(response: 
    CancelableRequest<Response>
    , schemas: S) {
        type FilterStartingWith<S extends PropertyKey, T extends string> = S extends number | string ? \`\${S}\` extends \`\${T}\${infer _X}\` ? S : never : never
        type InferSchemaType<T> = T extends { is: (o: unknown) => o is infer S } ? S : never
        const result = await response
        const schema = schemas[result.statusCode] ?? schemas.default
        const body = schema?.parse?.(result.body) ?? result.body
        return {statusCode: result.statusCode, headers: result.headers, body: body as InferSchemaType<S[keyof Pick<S, FilterStartingWith<keyof S, '2' | 'default'>>]> }
    }
}

",
    "examples/restclients/zod/xkcd/xkcd.zod.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import { z } from 'zod'

export const Comic = z.object({
    alt: z.string().optional(),
    day: z.string().optional(),
    img: z.string().optional(),
    link: z.string().optional(),
    month: z.string().optional(),
    news: z.string().optional(),
    num: z.number().optional(),
    safe_title: z.string().optional(),
    title: z.string().optional(),
    transcript: z.string().optional(),
    year: z.string().optional(),
})

export type Comic = z.infer<typeof Comic>

",
  },
}
`;
